### PR TITLE
Add support for Content Registry API

### PR DESCRIPTION
**Pull request by ivanetchart**
Implements #571

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,6 +30,7 @@ to add new and/or missing endpoints. Currently the following services are suppor
 - [x] Broadcast Messages
 - [x] Commits
 - [x] Custom Attributes
+- [x] Content Registry
```
Comment (svanharmelen) [inline README.md]: This should be `Container Registry`

File: gitlab.go [hunk 2]
```diff
@@ -332,6 +332,7 @@ type Client struct {
 	Projects              *ProjectsService
 	ProtectedBranches     *ProtectedBranchesService
 	ProtectedTags         *ProtectedTagsService
+	Registry              *RegistryService
```
Comment (svanharmelen) [inline gitlab.go]: This should be `ContainerRegistry`

File: registry.go [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+	"fmt"
+	"net/url"
+	"time"
+)
+
+// RegistryService handles communication with the container registry related methods
+// of the GitLab API.
+//
+// GitLab API docs: https://docs.gitlab.com/ee/api/container_registry.html
+type RegistryService struct {
```
Comment (svanharmelen) [inline registry.go]: This should be `ContainerRegistryService`

All these name suggestions are just to keep the naming conventions in this package consistent.

File: registry.go [hunk 4]
```diff
@@ -0,0 +1,220 @@
+package gitlab
+
+import (
+	"fmt"
+	"net/url"
+	"time"
+)
+
+// RegistryService handles communication with the container registry related methods
```
Comment (svanharmelen) [inline registry.go]: Just as a FYI... I try to keep comments below 80chars and LoC below 100chars. So this comment would then become:

```
// RegistryService handles communication with the container registry related
// methods of the Gitlab API.
```

A valid exception for comments is when a URL is longer the 80chars and is already on its own line and a valid exception for LoC is function signatures, as I always let them be on a single line. Thanks!

File: registry.go [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+
+// RegistryRepository represents a GitLab content registry repository.
+//
+// GitLab API docs: https://docs.gitlab.com/ee/api/container_registry.html
+type RegistryRepository struct {
+	ID        int        `json:"id"`
+	Name      string     `json:"name"`
+	Path      string     `json:"path"`
+	Location  string     `json:"location"`
+	CreatedAt *time.Time `json:"created_at,omitempty"`
```
Comment (svanharmelen) [inline registry.go]: `omitempty` can be omitted here as this struct is only used to unmarshal the response.

File: registry.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+type RegistryRepositoryTagDetail struct {
+	Name          string     `json:"name"`
+	Path          string     `json:"path"`
+	Location      string     `json:"location"`
+	Revision      string     `json:"revision"`
+	ShortRevision string     `json:"short_revision"`
+	Digest        string     `json:"digest"`
+	CreatedAt     *time.Time `json:"created_at,omitempty"`
+	TotalSize     int        `json:"total_size"`
+}
```
Comment (svanharmelen) [inline registry.go]: I wonder if we should use two different structs for `RegistryRepositoryTag` and `RegistryRepositoryTagDetail`? I would suggest to just use on struct instead. What do you think?

File: registry.go [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+// BulkDeleteRegistryRepositoryTagsOptions represents the available BulkDeleteRegistryRepositoryTags() options.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/container_registry.html#delete-repository-tags-in-bulk
+type BulkDeleteRegistryRepositoryTagsOptions struct {
+	NameRegexp *string `url:"name_regex,omitempty" json:"name_regex,omitempty"`
+	KeepN      *int    `url:"keep_n,omitempty" json:"keep_n,omitempty"`
+	OlderThan  *string `url:"older_than,omitempty" json:"older_than,omitempty"`
+}
```
Comment (svanharmelen) [inline registry.go]: Please define these option structs just above the first function that requires them.

File: registry.go [hunk 8]
```diff
@@ -0,0 +1,220 @@
...
+	}
+
+	return repos, resp, err
+}
+
+// DeleteRegistryRepository deletes a repository in a registry.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/container_registry.html#delete-registry-repository
+func (s *RegistryService) DeleteRegistryRepository(pid interface{}, repositoryID int, options ...OptionFunc) (*Response, error) {
```
Comment (svanharmelen) [inline registry.go]: To keep the naming consistent, `repositoryID` should be `repository`. Goes for all other methods as well of course.

File: registry.go [hunk 9]
```diff
@@ -0,0 +1,220 @@
...
+	project, err := parseID(pid)
+	if err != nil {
+		return nil, err
+	}
+
```
Comment (svanharmelen) [inline registry.go]: Please remove this blank line (again for consistency 😉) Goes for other methods as well of course.

File: registry.go [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+	u := fmt.Sprintf("projects/%s/registry/repositories/%d/tags", url.QueryEscape(project), repositoryID)
+
+	req, err := s.client.NewRequest("GET", u, opt, options)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var repos []*RegistryRepositoryTag
```
Comment (svanharmelen) [inline registry.go]: `repos` should probably be `tags` here?

File: registry.go [hunk 11]
```diff
@@ -0,0 +1,220 @@
...
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/container_registry.html#get-details-of-a-repository-tag
+func (s *RegistryService) GetRegistryRepositoryTagDetail(pid interface{}, repositoryID int, tagName string, options ...OptionFunc) (*RegistryRepositoryTagDetail, *Response, error) {
+	project, err := parseID(pid)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	u := fmt.Sprintf("projects/%s/registry/repositories/%d/tags/%s", url.QueryEscape(project), repositoryID, tagName)
```
Comment (svanharmelen) [inline registry.go]:
```suggestion
	u := fmt.Sprintf("projects/%s/registry/repositories/%d/tags/%s",
		url.QueryEscape(project),
		repositoryID,
		tagName,
	)
```

File: registry.go [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+
+	req, err := s.client.NewRequest("GET", u, nil, options)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	var tagDetail *RegistryRepositoryTagDetail
```
Comment (svanharmelen) [inline registry.go]:
```suggestion
	td := new(RegistryRepositoryTagDetail)
```

Comment (svanharmelen) [reply registry.go]: Again, just to keep things consistent ^^

File: registry.go [hunk 13]
```diff
@@ -0,0 +1,220 @@
...
+	}
+
+	return s.client.Do(req, nil)
+}
+
+// BulkDeleteRegistryRepositoryTags deletes repository tags in bulk based on given criteria.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/container_registry.html#delete-repository-tags-in-bulk
+func (s *RegistryService) BulkDeleteRegistryRepositoryTags(pid interface{}, repositoryID int, opt *BulkDeleteRegistryRepositoryTagsOptions, options ...OptionFunc) (*Response, error) {
```
Comment (svanharmelen) [inline registry.go]: I would rename this method to `DeleteRegistryRepositoryTags` to make it just a bit shorter.

File: registry.go [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/container_registry.html#delete-a-repository-tag
+func (s *RegistryService) DeleteRegistryRepositoryTag(pid interface{}, repositoryID int, tagName string, options ...OptionFunc) (*Response, error) {
+	project, err := parseID(pid)
+	if err != nil {
+		return nil, err
+	}
+
+	u := fmt.Sprintf("projects/%s/registry/repositories/%d/tags/%s", url.QueryEscape(project), repositoryID, tagName)
```
Comment (svanharmelen) [inline registry.go]:
```suggestion
	u := fmt.Sprintf("projects/%s/registry/repositories/%d/tags/%s",
		url.QueryEscape(project),
		repositoryID,
		tagName,
	)
```

File: registry.go [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/container_registry.html#list-repository-tags
+type ListRegistryRepositoryTagsOptions ListOptions
+
+// BulkDeleteRegistryRepositoryTagsOptions represents the available BulkDeleteRegistryRepositoryTags() options.
+//
+// GitLab API docs:
+// https://docs.gitlab.com/ee/api/container_registry.html#delete-repository-tags-in-bulk
+type BulkDeleteRegistryRepositoryTagsOptions struct {
```
Comment (svanharmelen) [inline registry.go]: I would call this `DeleteRegistryRepositoryTagsOptions` to make it just a bit shorter. I understand that the methods names would them be `DeleteRegistryRepositoryTag` vs `DeleteRegistryRepositoryTags` which is very similar, but still clear enough I think.

File: README.md (hunk 1 above)
Comment (ivanetchart) [reply README.md]: Totally, I don't know why I used `Content` 😕

File: registry.go (hunk 6 above)
Comment (ivanetchart) [reply registry.go]: I think it makes sense. `RegistryRepositoryTag` is a subset of `
RegistryRepositoryTagDetail`. 👍

**Comment by ivanetchart** [comment]
@svanharmelen I think most of the changes should be covered. I need to test a little bit more the bulk delete operation as it's done asynchronously and obviously need a bigger amount of images to make sure it's actually working.

**Comment by svanharmelen** [comment]
LGTM! I'm going to merge this in as is now. If you need to make tweaks to the bulk methods I prefer those in a new smaller PR 😉 Thanks!

**Pull request merged**